### PR TITLE
[WIP] Fixes for dask.order - Remove change of tactical goal in single dep path

### PR DESCRIPTION
**Pull request by fjetter**
This is waaaay too early for a PR but considering how complex this subject is I wanted to share intermediate progress asap.

This is specifically looking at the graph taken from https://github.com/dask/dask/issues/10384. The simplified reproducer I provided over there is part of this test and I started investigating why this is ordered poorly

test_xarray_reduction

## Main ordering

test_xarray_reduction_order-age

Pressure 8

## Optimal ordering

I ordered the graph manually to what I consider optimal ordering for this case. I don't necessarily strive to reproduce this but this should serve as an orientation. This is likely not unique. Note how source nodes have a homogeneous age and how terminal/end nodes are computed more quickly.

test_xarray_reduction_order-age-optimal

Pressure 6

## This PR (WIP)

As a first step, I looked into why `("b", 1, 0)` is scheduled way too late with priority 22. It should be scheduled immediately after `("a", 1, 0)` to free up `("a", 0, 0)`.

It turns out that this task is scheduled poorly due to a "change of the tactical goal" that is based on the partition_keys. Removing this entirely "fixes" this specific section of the graph and allows `("b", 1, 0)` to be scheduled much more quickly, reducing pressure from 8 to 6.

That's already nice but the resulting ordering is still far away from optimal and a little to heterogenous for my taste considering how symmetrical this problem is.

I haven't found out, yet, whether this "change of tactical goal" is a bad idea in general or whether the partition_key cost function is just bad. FWIW this specific code branch is untested. While it is _covered_ by `test_map_overlap` and `test_array_store_final_order`, removing this condition entirely does not have an impact on any asserted property. In fact, at least for `test_map_overlap`, removing this condition also reduces pressure and generates a better graph (the other graph is more difficult to analyze)

test_xarray_reduction_order-age

cc @eriknw

### CODE REVIEW

File: dask/order.py [hunk 1]
```diff
@@ -363,8 +363,8 @@ def finish_now_key(x):
                         else:
                             inner_stack.extend(deps)
                         seen_update(deps)
-                    if not singles:
-                        continue
+                    # if not singles:
+                    continue
                     process_singles = True
```
Comment (fjetter) [inline dask/order.py:368]: Disabling this JUMP now actually generates the optimal order

**Comment by fjetter** [comment]
I found two smart switches that actually cause this specific graph to be ordered worse. However, the linear growth described in https://github.com/dask/dask/issues/10384#issuecomment-1715841811 is still a problem.

File: dask/order.py (hunk 1 above)
Comment (fjetter) [reply dask/order.py:368]: however, this also breaks some tests now. Have to investigate further

**Comment by fjetter** [comment]
Note: I'm just documenting observations here. There are no deep insights here and I only recommend reading this if you are *really* interested and want to following along

Considering that https://github.com/dask/dask/issues/10384#issuecomment-1715841811 shows that dataframes scale *much* better, I started looking into what the difference between the two graphs is.

- both graphs share the same keys. In fact, the array graph is a true subset of the DataFrame graph (`set(mean.__dask_graph__()).issubset(set(mean.to_dask_dataframe().__dask_graph__())) is True`)
- Both graphs start ordering with the same key

After the initial pick of the key, the two orderings diverge already starting with the pick of the second key. Effectively, the first key is a random number gen that breaks into a `pow` and a `mul` task and the decision about which one to pick next is done [here](https://github.com/dask/dask/blob/3520077f95c790b0413235a813b1acd491d7b533/dask/order.py#L590-L599) by relying on the _partition_key_ that is defined as

https://github.com/dask/dask/blob/3520077f95c790b0413235a813b1acd491d7b533/dask/order.py#L221-L233

This partition_keys function is also used to "change the strategic goal" that I disabled earlier. Inspecting the two partition keys for both the dataframe and array workloads reveal an interesting property

| Task  | Array | DataFrame
| ------------- | ------------- |-|
| `pow`  | 84  | 300 |
| `mul`  | 108  | 299 |

So, not only does the dataframe choose a different path, the partition_keys are vastly different in both cases even though the graphs are sufficiently similar (I'll get to the topological differences in a moment)


Inspecting why that is, we look at the calculated graph metrics


| `pow`  | Array | DataFrame
| ------------- | ------------- |-|
| `total_dependents`  | 5  | 8 |
| `min_dependencies`  | 85  | 301 |
| `max_dependencies`  | 85  | 301 |
| `min_height`  | 4  | 7 |
| `total_dependencies`  | 2  | 2 |


| `mul`  | Array | DataFrame
| ------------- | ------------- |-|
| `total_dependents`  | 5  | 8 |
| `min_dependencies`  | 110  | 301 |
| `max_dependencies`  | 110  | 301 |
| `min_height`  | 4  | 7 |
| `max_height`  | 4  | 7 |
| `total_dependencies`  | 3  | 3 |

So, all numbers are very different. The partition_key is a function of most of those metrics `partition_key(min_dependencies, total_dependents, min_heights, total_dependencies)`

In a very, very, very simplified version of this graph, we can see that the `mul` tasks require an additional dependency while the `pow` tasks could proceed further immediately

image


Now, in terms of the overall differences between the graph, the two versions (very small) below
# DataFrame
dataframe-small
# Array
array-small

The most obvious difference between arrays and DataFrames is that the DataFrame has a single final task as output key while the array graph has three (the three sub-arrays of the xarray datasets). This may be relevant since dask.order is actually implementing a special case for this scenario https://github.com/dask/dask/blob/3520077f95c790b0413235a813b1acd491d7b533/dask/order.py#L128-L135 and likely also explains the significant differences in the metrics, particularly `min_dependencies`

to be continued...

**Comment by fjetter** [comment]
I did a little time traveling and measured pressure for dataframe / array

9983c19fd9770d1a5cee2ef8b63024a934a626ee 51 / 511
439c4ab28cd3ed3de23c13cbbe3e0db8342dd4de 1011 / 1011
816cfa757e5239c7284040c8f1f9ae7f8a6b38cf 1011 / 1011
c63dc149210d45cbf90bb3f9d5a56106339ac24c 514 / 513
674149506fb28e566106aa6cebb14a27001b35e7 348 / 205
59d8c9c09199afefd16de8503de1fa7323cd2eba 348 / 348
f4a753129f9cd8e290373ce1ddf926adfe18ead7 135 / 135
f5779130fc2a733cc3c3ad128124d88d0292190c 1011 / 1011
11a50f0d329bdaf1ea6b7f0cff9500f55699fd36 1011 / 1011

The good news is that we apparently are not at the worst case (which seems to be 1011 for this example) and this is no clear regression. However, it looks like the version of f4a753129f9cd8e290373ce1ddf926adfe18ead7 was doing pretty well for both cases (and it was symmetrical)

**Comment by mrocklin** [comment]
I'd be curious what this would look like applied to the A/B tests.  Presumably recent fixes were for some reason.  It seems like we weren't measuring well though, except with unit tests.  Applying these to a broader set of workflows would be interesting.

**Comment by fjetter** [comment]
> I'd be curious what this would look like applied to the A/B tests.

Running. This will take a moment

**Comment by fjetter** [comment]
Just looking over the version of 0.17.2 though, I doubt it is sufficient. There are many relatively obvious failure cases it does not handle. However, it is actually more robust for the cases I'm currently looking at (even _fixes_ https://github.com/dask/dask/issues/9995)

**Pull request closed without merging**